### PR TITLE
Track tab routes

### LINKED ISSUE #1185: Track tab routes in Firebase Analytics
Track views of the pages in the main tab router

---

**Please follow this checklist. Please check each appropriate box (put an 'x' or check it after creating the issue).**

- [x] Searched the [existing issues](https://github.com/WorldHealthOrganization/app/issues) to ensure you are not creating a duplicate.
- [x] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md).
- [x] Issue has descriptive title focused on the feature or problem from the user's perspective

### PR DESCRIPTION
**Pull request by jasonTelanoff**
Added Tracking of Tab Bars

Closes #1185

---

#### Please follow this checklist. Please check each appropriate box (put an 'x' or check it after creating the PR).

- [x] REQUIRED: Do you have an Issue **assigned to you** for this PR?
- [x] Provided detailed pull request description and a succinct title
- [x] Tested your changes, especially after any code review iterations.
- [x] Included any relevant screenshots of UI updates.
- [x] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md).
- [x] Verified all contributions are properly licensed pursuant to the [LICENSE](https://github.com/WorldHealthOrganization/app/blob/master/LICENSE) file in the root of the repository.
- [x] Verified your name is in the [content/credits.yaml](https://github.com/WorldHealthOrganization/app/blob/master/content/credits.yaml) file (if you want it to be).

### CODE REVIEW

File: client/flutter/lib/pages/main_pages/app_tab_router.dart [hunk 1]
```diff
@@ -40,6 +41,52 @@ class AppTabRouter extends StatelessWidget {
 
   AppTabRouter(this.tabs, this.navItems);
 
+  @override
+  _AppTabRouterState createState() => _AppTabRouterState(tabs, navItems);
```
Comment (hspinks) [inline client/flutter/lib/pages/main_pages/app_tab_router.dart:45]: Instead of passing params, use `this.widget.tabs` and `this.widget.navItems` in the state class. See [this answer](https://stackoverflow.com/a/50818870/2434476) for detail

```suggestion
  _AppTabRouterState createState() => _AppTabRouterState();
```

File: client/flutter/lib/pages/main_pages/app_tab_router.dart [hunk 2]
```diff
@@ -40,6 +41,52 @@ class AppTabRouter extends StatelessWidget {
...
+  String tabTitle(int index) {
+    Text txt = AppTabRouter.defaultNavItems[index].title;
+    String data = txt.data;
+    return data;
+  }
+
+  final List<Widget Function(BuildContext)> tabs;
+  final List<BottomNavigationBarItem> navItems;
+
+  _AppTabRouterState(this.tabs, this.navItems);
```
Comment (hspinks) [inline client/flutter/lib/pages/main_pages/app_tab_router.dart:88]: Same comment as above - access these vars through the widget, if they're needed at all

```suggestion
  const _AppTabRouterState();
```

File: client/flutter/lib/pages/main_pages/app_tab_router.dart [hunk 3]
```diff
@@ -40,6 +41,52 @@ class AppTabRouter extends StatelessWidget {
...
+  CupertinoTabController _controller;
+  FirebaseAnalytics _analytics;
+  int lastTabIndex;
+  @override
+  void initState() {
+    _analytics = FirebaseAnalytics();
+    _controller = CupertinoTabController();
+    lastTabIndex = _controller.index;
+
+    _controller.addListener(
```
Comment (hspinks) [inline client/flutter/lib/pages/main_pages/app_tab_router.dart:58]: pardon my unfamiliarity with Dart, but does there need to be any cleanup of listeners?

File: client/flutter/lib/pages/main_pages/app_tab_router.dart [hunk 4]
```diff
@@ -40,6 +41,52 @@ class AppTabRouter extends StatelessWidget {
...
+        _analytics.logEvent(
+          name: 'changed_tabs',
+          parameters: {
+            'current_tab': currentTab,
+            'last_tab': lastTab,
+          },
+        );
+      },
+    );
+    super.initState();
```
Comment (hspinks) [inline client/flutter/lib/pages/main_pages/app_tab_router.dart:76]: It's generally good practice to put your superclass calls to the method you're overriding at the beginning of your method, just in case the superclass changes things you're setting up here or initializes things required for your code. Please move the `super.initState();` to the first line in this method

File: client/flutter/lib/pages/main_pages/app_tab_router.dart [hunk 5]
```diff
@@ -40,6 +41,52 @@ class AppTabRouter extends StatelessWidget {
...
+    _controller.addListener(
+      () {
+        int currentIndex = _controller.index;
+
+        String lastTab = tabTitle(lastTabIndex);
+        String currentTab = tabTitle(currentIndex);
+
+        lastTabIndex = currentIndex;
+
+        _analytics.logEvent(
```
[the remark this answers is not in the document]
Comment (advayDev1) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart:67]: I agree, but IMO this is a blocker: please make it a screen_view and match the schema.  This was discussed in slack.
Almost all other analytics screen events were converted from random events to screen_views with named roting.

[the remark this answers is not in the document]
Comment (jasonTelanoff) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart:62]: sure thing

File: client/flutter/lib/pages/main_pages/app_tab_router.dart (hunk 4 above)
Comment (jasonTelanoff) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart]: done

File: client/flutter/lib/pages/main_pages/app_tab_router.dart (hunk 3 above)
Comment (jasonTelanoff) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart:64]: I do not believe so

File: client/flutter/lib/pages/main_pages/app_tab_router.dart [hunk 6]
```diff
@@ -20,25 +26,66 @@ class AppTabRouter extends StatelessWidget {
     (context) => SettingsPage(),
   ];
 
-  static final List<BottomNavigationBarItem> defaultNavItems = [
+  final List<BottomNavigationBarItem> navItems = [
```
Comment (advayDev1) [inline client/flutter/lib/pages/main_pages/app_tab_router.dart:29]: why did you remove the default?

File: client/flutter/lib/pages/main_pages/routes.dart [hunk 7]
```diff
@@ -16,8 +16,7 @@ import 'package:who_app/pages/travel_advice.dart';
 
 class Routes {
   static final Map<String, WidgetBuilder> map = {
-    '/': (context) =>
-        AppTabRouter(AppTabRouter.defaultTabs, AppTabRouter.defaultNavItems),
+    '/': (context) => AppTabRouter(),
```
Comment (advayDev1) [inline client/flutter/lib/pages/main_pages/routes.dart:19]: why did you revert this change?

File: client/flutter/lib/pages/main_pages/app_tab_router.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (hspinks) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart:73]: @jasonTelanoff sorry for the delay here, here's an example: https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_analytics/firebase_analytics/example/lib/tabs_page.dart#L91

[the remark this answers is not in the document]
Comment (hspinks) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart:73]: Should be able to use the [`setCurrentScreen` method](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart#L91)

[the remark this answers is not in the document]
Comment (jasonTelanoff) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart]: Cool, I just pushed the code

File: client/flutter/lib/pages/main_pages/routes.dart (hunk 7 above)
Comment (theswerd) [reply client/flutter/lib/pages/main_pages/routes.dart]: Misunderstood comments above -- just pushed fix

File: client/flutter/lib/pages/main_pages/app_tab_router.dart (hunk 6 above)
Comment (jasonTelanoff) [reply client/flutter/lib/pages/main_pages/app_tab_router.dart]: this has been reverted

**Comment by theswerd** [comment]
@hspinks I can confirm it works

**Comment by jasonTelanoff** [comment]
I did some tests, but you can never be sure enough. Are there any other issues?

**Comment by advayDev1** [comment]
Hey @jasonTelanoff thanks for your work on this PR.  There is a conflict to resolve, but please check out Slack announcements as well.

Feel free to DM me in Slack if you have any questions
